drivers/rn2xx3/rn2xx3.c

```diff
@@ -0,0 +1,303 @@
+    snprintf(dev->cmd_buf, 20, "sys sleep %lu", dev->sleep);
+    if (rn2xx3_write_cmd_no_wait(dev) == RN2XX3_ERR_INVALID_PARAM) {
+        DEBUG("[rn2xx3] sleep: cannot put module in sleep mode\n");
+        return RN2XX3_ERR_INVALID_PARAM;
+    }
+
+    rn2xx3_set_internal_state(dev, RN2XX3_INT_STATE_SLEEP);
+    xtimer_set(&dev->sleep_timer, dev->sleep * US_PER_MS);
+
+    return RN2XX3_OK;
+}
+
+int rn2xx3_mac_init(rn2xx3_t *dev)
+{
+    rn2xx3_mac_set_dev_eui(dev, dev->p.loramac.dev_eui);
+    rn2xx3_mac_set_app_eui(dev, dev->p.loramac.app_eui);
+#ifdef LORAMAC_APP_APPKEY
+    rn2xx3_mac_set_app_key(dev, dev->p.loramac.app_key);
+#endif
+#ifdef LORAMAC_APP_SESSION_APPKEY
+    rn2xx3_mac_set_apps_key(dev, dev->p.loramac.app_skey);
+#endif
+#ifdef LORAMAC_NET_SESSION_KEY
+    rn2xx3_mac_set_nwkskey_key(dev, dev->p.loramac.net_skey);
+#endif
+
+    rn2xx3_mac_set_dr(dev, LORAMAC_DEFAULT_DR);
+    rn2xx3_mac_set_tx_power(dev, LORAMAC_DEFAULT_TX_POWER);
+    rn2xx3_mac_set_tx_port(dev, LORAMAC_DEFAULT_TX_PORT);
+    rn2xx3_mac_set_tx_mode(dev, LORAMAC_DEFAULT_TX_MODE);
+    rn2xx3_mac_set_adr(dev, LORAMAC_DEFAULT_ADR);
+    rn2xx3_mac_set_retx(dev, LORAMAC_DEFAULT_RETX);
+    rn2xx3_mac_set_linkchk_interval(dev, LORAMAC_DEFAULT_LINKCHK);
+    rn2xx3_mac_set_rx1_delay(dev, LORAMAC_DEFAULT_RX1_DELAY);
+    rn2xx3_mac_set_ar(dev, LORAMAC_DEFAULT_AR);
+    rn2xx3_mac_set_rx2_dr(dev, LORAMAC_DEFAULT_RX2_DR);
+    rn2xx3_mac_set_rx2_freq(dev, LORAMAC_DEFAULT_RX2_FREQ);
+
+    if (LORAMAC_DEFAULT_JOIN_PROCEDURE == LORAMAC_JOIN_ABP) {
+        DEBUG("[rn2xx3] setting device addr only in ABP mode\n");
+        rn2xx3_mac_set_dev_addr(dev, dev->p.loramac.dev_addr);
+        /* Store parameters to EEPROM */
+        rn2xx3_mac_save(dev);
+    }
+
+    rn2xx3_sys_reset(dev);
+
+    return RN2XX3_OK;
+}
+
+int rn2xx3_mac_tx(rn2xx3_t *dev, uint8_t *payload, uint8_t payload_len)
+{
+    if (dev->int_state == RN2XX3_INT_STATE_SLEEP) {
+        DEBUG("[rn2xx3] ABORT: device is in sleep mode\n");
+        return RN2XX3_ERR_SLEEP_MODE;
+    }
+
+    rn2xx3_mac_tx_start(dev);
+    for (unsigned i = 0; i < payload_len; ++i) {
+        rn2xx3_cmd_append(dev, &payload[i], 1);
+    }
+
+    int ret = rn2xx3_mac_tx_finalize(dev);
+    if (ret != RN2XX3_OK) {
+        DEBUG("[rn2xx3] TX command failed\n");
+        return ret;
+    }
+
+    rn2xx3_set_internal_state(dev, RN2XX3_INT_STATE_MAC_TX);
+
+    ret = rn2xx3_wait_reply(dev, RN2XX3_REPLY_DELAY_TIMEOUT);
+
+    return ret;
+}
+
+int rn2xx3_mac_join_network(rn2xx3_t *dev, uint8_t mode)
+{
+    memset(dev->cmd_buf, 0, strlen(dev->cmd_buf));
+    snprintf(dev->cmd_buf, 10, "mac join ");
+    switch (mode) {
+        case LORAMAC_JOIN_OTAA:
+            strcat(dev->cmd_buf, "otaa");
+            break;
+
+        case LORAMAC_JOIN_ABP:
+            strcat(dev->cmd_buf, "abp");
+            break;
+
+        default:
+            DEBUG("[rn2xx3] unsupported join procedure mode\n");
+            break;
```

Comment:
No return. We're going to send "mac join " and let the device come up with an error ?